rectools/models/sasrec.py

```diff
@@ -922,57 +928,108 @@ def on_train_start(self) -> None:
+        pos_probs_adjusted = torch.clamp(pos_probs.pow(-beta), 1 + epsilon, torch.finfo(torch.float64).max)
+        pos_probs_adjusted = torch.clamp(
+            torch.div(1, (pos_probs_adjusted - 1)), epsilon, torch.finfo(torch.float64).max
+        )
+        pos_logits_transformed = torch.log(pos_probs_adjusted)
+        logits = torch.cat([pos_logits_transformed, neg_logits], dim=-1)
+        return logits
 
-            # Logits must be passed in form of [batch_size, n_items + n_special_tokens, session_max_len],
-            #  where n_items + n_special_tokens is number of classes
+    @classmethod
+    def _calc_softmax_loss(cls, logits: torch.Tensor, y: torch.Tensor, w: torch.Tensor) -> torch.Tensor:
+        # We are using CrossEntropyLoss with a multi-dimensional case
 
-            # Target label indexes must be passed in a form of [batch_size, session_max_len]
-            # (`0` index for "PAD" ix excluded from loss)
+        # Logits must be passed in form of [batch_size, n_items + n_item_extra_tokens, session_max_len],
+        #  where n_items + n_item_extra_tokens is number of classes
 
-            # Loss output will have a shape of [batch_size, session_max_len]
-            # and will have zeros for every `0` target label
+        # Target label indexes must be passed in a form of [batch_size, session_max_len]
+        # (`0` index for "PAD" ix excluded from loss)
 
-            loss = torch.nn.functional.cross_entropy(
-                logits.transpose(1, 2), y, ignore_index=0, reduction="none"
-            )  # [batch_size, session_max_len]
-            loss = loss * w
-            n = (loss > 0).to(loss.dtype)
-            loss = torch.sum(loss) / torch.sum(n)
-            return loss
-        raise ValueError(f"loss {loss} is not supported")
+        # Loss output will have a shape of [batch_size, session_max_len]
+        # and will have zeros for every `0` target label
+        loss = torch.nn.functional.cross_entropy(
+            logits.transpose(1, 2), y, ignore_index=0, reduction="none"
+        )  # [batch_size, session_max_len]
+        loss = loss * w
+        n = (loss > 0).to(loss.dtype)
+        loss = torch.sum(loss) / torch.sum(n)
+        return loss
+
+    @classmethod
+    def _calc_bce_loss(cls, logits: torch.Tensor, y: torch.Tensor, w: torch.Tensor) -> torch.Tensor:
+        mask = y != 0
+        target = torch.zeros_like(logits)
+        target[:, :, 0] = 1
+
+        loss = torch.nn.functional.binary_cross_entropy_with_logits(
+            logits, target, reduction="none"
+        )  # [batch_size, session_max_len, n_negatives + 1]
+        loss = loss.mean(-1) * mask * w  # [batch_size, session_max_len]
+        loss = torch.sum(loss) / torch.sum(mask)
+        return loss
+
+    def _calc_gbce_loss(
+        self, logits: torch.Tensor, y: torch.Tensor, w: torch.Tensor, negatives: torch.Tensor
+    ) -> torch.Tensor:
+        n_items = self.torch_model.item_model.n_items
```

Comment:
```suggestion
        n_actual_items = self.torch_model.item_model.n_items - self.n_item_extra_tokens
```
Let's keep all logic in one place to simplify understanding. (I really don't like that we call `n_items` for two different things, it's bad. But for now can't fix it everywhere)

Reply:
fixed